pymeasure/instruments/thorlabs/thorlabspm100usb.py

```diff
@@ -23,74 +23,97 @@
-        if wavelength < self.wavelength_min:
-            raise RangeException("Wavelength %.2f nm out of range: using minimum wavelength: %.2f nm" % (
-                wavelength, self.wavelength_min))
-            # explicit setting wavelenghth, althought it would be automatically set
-            wavelength = self.wavelength_min
-        if wavelength > self.wavelength_max:
-            raise RangeException("Wavelength %.2f nm out of range: using maximum wavelength: %.2f nm" % (
-                wavelength, self.wavelength_max))
-            wavelength = self.wavelength_max
-        self.wavelength = wavelength
-        return self.power
-
-    def sensor(self):
-        "Get sensor info"
-        response = self.ask("SYST:SENSOR:IDN?").split(',')
+        self._set_flags()
+
+    wavelength_min = Instrument.measurement(
+        "SENS:CORR:WAV? MIN", "Minimum wavelength, in nm"
+    )
+
+    wavelength_max = Instrument.measurement(
+        "SENS:CORR:WAV? MAX", "Maximum wavelength, in nm"
+    )
+
+    @property
+    def wavelength(self):
+        """Wavelength in nm."""
+        value = self.values("SENSE:CORR:WAV?")[0]
+        return value
+
+    @wavelength.setter
+    def wavelength(self, value):
+        """Wavelength in nm."""
+        if self.wavelength_settable:
+            value = truncated_range(value, [self.wavelength_min, self.wavelength_max])
+            self.write("SENSE:CORR:WAV {}".format(value))
+        else:
+            raise AttributeError(
+                f"{self.sensor_name} does not allow setting the wavelength."
+            )
+
+    @property
+    def power(self):
+        """Power in W."""
+        if self.is_power:
+            return self.values("MEAS:POW?")[0]
+        else:
+            raise AttributeError(f"{self.sensor_name} is not a power sensor.")
+
+    @property
+    def energy(self):
+        """Energy in J."""
+        if self.is_energy:
+            return self.values("MEAS:ENER?")[0]
+        else:
+            raise AttributeError(f"{self.sensor_name} is not an energy sensor.")
+
+    def _set_flags(self):
+        """Get sensor info and write flags."""
+        response = self.ask("SYST:SENSOR:IDN?").split(",")
         self.sensor_name = response[0]
         self.sensor_sn = response[1]
         self.sensor_cal_msg = response[2]
         self.sensor_type = response[3]
         self.sensor_subtype = response[4]
-        self._flags_str = response[-1][:-1]
-        # interpretation of the flags
-        # rough trick using bin repr, maybe something more elegant exixts
-        # (bitshift, bitarray?)
-        self._flags = tuple(
-            map(lambda x: x == '1', bin(int(self._flags_str))[2:]))
-        # setting the flags; _dn are empty
-        self.is_power, self.is_energy, _d4, _d8, \
-        self.resp_settable, self.wavelength_settable, self.tau_settable, _d128, self.temperature_sens = self._flags
+        _flags_str = response[5].rstrip("\n")
 
```

Comment:
I think the `rstrip` is not needed if you use `self.values` above or you need to preserve trailing spaces?

Reply:
Changed it in 27bb492.